components/layout/construct.rs

```diff
@@ -705,9 +705,16 @@ impl<'a, ConcreteThreadSafeLayoutNode: ThreadSafeLayoutNode>
         // List of absolute descendants, in tree order.
         let mut abs_descendants = AbsoluteDescendants::new();
         let mut legalizer = Legalizer::new();
-        if !node.is_replaced_content() {
+        let is_media_element =
+            node.type_id() == Some(LayoutNodeType::Element(LayoutElementType::HTMLMediaElement));
+        if !node.is_replaced_content() || is_media_element {
             for kid in node.children() {
-                if kid.get_pseudo_element_type() != PseudoElementType::Normal {
+                if is_media_element && !kid.is_user_agent_widget() {
```

Comment:
Why is this needed? How can a non-UA-widget thing end up here?

Reply:
Without this check, something like:

```
<audio><img src="image.jpg"></audio>
```

would render the image.

Reply:
Sure, then you need to do:

```
let is_media_element_with_widget = node.type_id() == ... && node.as_element().unwrap().shadow_root().is_some();
```

Right? _That_ is the case we want to enter this loop, not the rest.

Reply:
The media element is not a shadow root in this case. It's the parent of a shadow root. That's is why I was adding the check inside the `node.children()` loop.

Reply:
It is true that checking the existance of the shadow root is enough, as we control that the only case where a shadow root can be attached to a media element is to serve as a UA widget.